munit/shared/src/main/scala/munit/FunFixtures.scala

```diff
@@ -1,47 +1,105 @@
 package munit
 
+import munit.internal.FutureCompat._
+
+import scala.concurrent.Future
+import scala.util.Success
+import scala.util.Failure
+
 trait FunFixtures { self: FunSuite =>
 
-  class FunFixture[T](
-      val setup: TestOptions => T,
-      val teardown: T => Unit
+  class FunFixture[T] private (
+      val setupAndTeardown: (TestOptions => Future[T], T => Future[Unit])
   ) {
+    def setup = setupAndTeardown._1
+    def teardown = setupAndTeardown._2
+
+    @deprecated("Use `FunFixture(...)` without `new` instead", "0.7.2")
+    def this(setup: TestOptions => T, teardown: T => Unit) =
+      this(
+        (
+          options => Future(setup(options))(munitExecutionContext),
+          argument => Future(teardown(argument))(munitExecutionContext)
+        )
+      )
+
     def test(options: TestOptions)(
         body: T => Any
     )(implicit loc: Location): Unit = {
       self.test(options) {
-        val argument = setup(options)
-        try body(argument)
-        finally teardown(argument)
+        implicit val ec = munitExecutionContext
+        setup(options).flatMap { argument =>
+          munitValueTransform(body(argument))
+            .transformWithCompat(o =>
+              teardown(argument).transformCompat {
+                case Success(_)     => o
+                case f @ Failure(_) => f
```

Comment:
We discard a possible failure in `o` here in favor of the exception from the teardown method. We can aggregate them into a single `Failure` like this instead
```suggestion
                case teardownFailure @ Failure(teardownException) =>
                  o match {
                    case bodyFailure @ Failure(bodyException) =>
                      bodyException.addSuppressed(teardownException)
                      bodyFailure
                    case _ =>
                      teardownFailure
```

Reply:
Good point!